### ISSUE TITLE
Failure to Deploy default skydive.yaml - agents running but not ready

### ISSUE THREAD

**Issue by developeer** [skydive-project/skydive#1895]
deployed skydive on 2 master 2 worker cluster running kubernetes 1.14.1 and ubuntu 18.04.2.  
We deployed skydive.yaml from contrib/kubernetes

The application is deployed.  The analyzer is running 2/2 but the agents are running but not ready.
Once of the agents is in CrashLoopBackOff.  

error from agent.

ERROR    probes/probes.go:104 NewFlowProbeBundle    kube-worker7: Failed to create ovsnetflow probe: Agent.ovsnetflow probe depends on agent.ovsdb topology probe: agent.ovsnetflow probe can't start properly.


In kube-apiserver on master
E0702 17:29:23.804995       1 repair.go:149] the node port 30187 for service skydive-analyzer/default was assigned to multiple services; please recreate

Does the configuration need to be changed?  

Error From the Analyzer many times for each k8s object. 
Failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:skydive-service-account" cannot list resource "secrets" in API group "" at the cluster scope

│ 2019-07-02T19:01:22.412Z    ERROR    probes/probes.go:104 NewFlowProbeBundle    kube-worker6: Failed to create **ovssflow** probe: Agent.ovssflow probe depends on agent.ovsdb topology probe: agent.ovssflow probe can't start properly                                                                                                               │
│ 2019-07-02T19:01:22.412Z    INFO    probes/probes.go:106 NewFlowProbeBundle    kube-worker6: Not compiled with dpdk support, skipping it                                                                                                                                                                                                           │
│ 2019-07-02T19:01:22.412Z    ERROR    probes/probes.go:104 NewFlowProbeBundle    kube-worker6: Failed to create **ovsmirror** probe: Agent.ovssflow probe depends on agent.ovsdb topology probe: agent.ovssflow probe can't start properly                                                                                                              │
│ 2019-07-02T19:01:22.412Z    ERROR    probes/probes.go:104 NewFlowProbeBundle    kube-worker6: Failed to create **ovsnetflow** probe: Agent.ovsnetflow probe depends on agent.ovsdb topology probe: agent.ovsnetflow probe can't start properly

**Comment by safchain** [comment]
Hi, the error reported is not fatal, can you please paste the full agent log, thanks.

**Comment by ekaley** [comment]
hey,

here are the agent logs were seeing errors in.

Using kubernetes 1.14.1 with privileged mode enabled on the api server and kublets

```kubectl logs -f skydive-agent-6crkv
2019-07-08T20:06:27.942Z    INFO    agent/agent.go:41 glob..func1    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Skydive Agent 0.23.0-d30bded6e89e starting...
2019-07-08T20:06:27.943Z    INFO    http/server.go:104 (*Server).Listen    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Listening on socket 127.0.0.1:8081
2019-07-08T20:06:27.948Z    INFO    agent/probes.go:45 NewTopologyProbeBundleFromConfig    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Topology probes: [runc docker]
2019-07-08T20:06:27.948Z    INFO    probes/probes.go:58 NewFlowProbeBundle    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Flow probes: [pcapsocket ovssflow sflow gopacket dpdk ebpf ovsmirror ovsnetflow]
2019-07-08T20:06:27.949Z    ERROR    probes/probes.go:104 NewFlowProbeBundle    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Failed to create ovssflow probe: Agent.ovssflow probe depends on agent.ovsdb topology probe: agent.ovssflow probe can't start properly
2019-07-08T20:06:27.949Z    INFO    probes/probes.go:106 NewFlowProbeBundle    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Not compiled with dpdk support, skipping it
2019-07-08T20:06:27.949Z    ERROR    probes/probes.go:104 NewFlowProbeBundle    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Failed to create ovsmirror probe: Agent.ovssflow probe depends on agent.ovsdb topology probe: agent.ovssflow probe can't start properly
2019-07-08T20:06:27.949Z    ERROR    probes/probes.go:104 NewFlowProbeBundle    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Failed to create ovsnetflow probe: Agent.ovsnetflow probe depends on agent.ovsdb topology probe: agent.ovsnetflow probe can't start properly
2019-07-08T20:06:27.949Z    INFO    agent/agent.go:51 glob..func1    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Skydive Agent started
2019-07-08T20:06:27.951Z    INFO    websocket/client.go:492 (*Client).Connect    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Connecting to ws://10.100.200.237:8082/ws/agent/topology
2019-07-08T20:06:27.953Z    INFO    websocket/client.go:516 (*Client).Connect    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Connected to ws://10.100.200.237:8082/ws/agent/topology
2019-07-08T20:06:27.954Z    INFO    pod/forwarder.go:56 (*Forwarder).OnNewMaster    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Using 10.100.200.237:8082 as master of topology forwarder
2019-07-08T20:06:27.954Z    INFO    pod/forwarder.go:37 (*Forwarder).triggerResync    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Start a re-sync for 0a24b74e-12c1-4ad8-a430-f57ddd80c20c
2019-07-08T20:07:41.365Z    WARN    netlink/netlink.go:1109 (*NetNsProbe).onMessageAvailable    0a24b74e-12c1-4ad8-a430-f57ddd80c20c: Failed to get Routes: Link not found```

**Comment by developeer** [comment]
@ekaley will provide more logs on issue.

**Comment by ekaley** [comment]
what are the prereqs for the base OS to have for packages needed for skydive, do we need openvswitch installed on the base image? Ubuntu 18.04 or 16.04?

**Comment by ekaley** [comment]
```Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  2m25s                default-scheduler  Successfully assigned default/skydive-agent-bfrfw to minikube
  Normal   Pulling    2m24s                kubelet, minikube  pulling image "skydive/skydive"
  Normal   Pulled     2m11s                kubelet, minikube  Successfully pulled image "skydive/skydive"
  Normal   Created    2m9s                 kubelet, minikube  Created container
  Normal   Started    2m9s                 kubelet, minikube  Started container
  Warning  Unhealthy  19s (x11 over 119s)  kubelet, minikube  Readiness probe failed: Get http://10.240.18.180:8081/api/status: dial tcp 10.240.18.180:8081: connect: connection refused
  Warning  Unhealthy  15s (x10 over 105s)  kubelet, minikube  Liveness probe failed: Get http://10.240.18.180:8081/api/status: dial tcp 10.240.18.180:8081: connect: connection refused
``` 

this is the error from the kubectl describe on the agent pod
Nothing else is using or running on 8081 on this node, host or cluster

**Comment by safchain** [comment]
@ekaley The OVS errors are not fatal errors it shouldn't the problem. @hunchback any clue on this ?

**Comment by ekaley** [comment]
Ive tried minikube and a CFCR kubernetes cluster deployment, still get connection refused on both deployments when trying to get the api status.  Analyzer comes up but none of the agents show ready

**Comment by developeer** [comment]
we are using k9s to look at the logs.  

![k9s-pods](https://user-images.githubusercontent.com/9946626/61301525-0e73d400-a7b2-11e9-85af-f37f44f1e739.png)

the skydive agents crash and do not startup.  The logs are not useful.  (which I sent above).  This is a basic k8s cluster.  2 master 2 workers.  

![k9s-agent-describe](https://user-images.githubusercontent.com/9946626/61302113-3c0d4d00-a7b3-11e9-95f1-fe21fa5ae62d.png)

**Comment by safchain** [comment]
Without complete agent log it will be hard to help on this... As soon as you see 

```
2019-07-16T16:37:13.232+0200	INFO	agent/agent.go:41 glob..func1	myhost: Skydive Agent 0.23.0-724cd0546cd3 starting...
2019-07-16T16:37:13.233+0200	INFO	http/server.go:104 (*Server).Listen	myhost: Listening on socket 127.0.0.1:8081
```
in the log the agent should be available even if there is error in the logs. If the agent is not available or crash something should be present in the logs like a stack trace

**Comment by developeer** [comment]
We can't access logs once the pod goes into crashloop.  The only thing we can get is the describe.

**Comment by developeer** [comment]
![k9s-agent-log](https://user-images.githubusercontent.com/9946626/61305213-6d3c4c00-a7b8-11e9-9d20-f314c625a35c.png)


this is the log from agent pod before crashloop

log from analyzer

![analyzer-log](https://user-images.githubusercontent.com/9946626/61305405-c310f400-a7b8-11e9-9dd1-15cdbe6c618d.png)

**Comment by safchain** [comment]
Per the agent log it seems that the agent is properly stopped, no crash here. could it be the healthcheck that stop the agent ? could you try to remove it ?

**Comment by developeer** [comment]
remove the livenessProbe from agent yaml and redeploy?

**Comment by safchain** [comment]
yes

**Comment by developeer** [comment]
Changed the yaml and redeployed.  This is the log.

![k9s-agent-logs-2](https://user-images.githubusercontent.com/9946626/61308828-10905f80-a7bf-11e9-8633-7201dc47e793.png)

**Comment by developeer** [comment]
We have installed the default skydive.yaml from the site on two different clusters with the same result.  I have given all of the logs possible.  The agents never get to ready state and crashbackupoff after several minutes.  We are running k8s 1.14.4 in a 2 master 2 worker cluster on vsphere.  It is a default setup of kubernetes.  Nothing special.

**Comment by safchain** [comment]
The last log report that the agent is started and connected to the analyzer. Without the livenessProbe thing you should have it reported as "started" at least for few minutes if it crash after awhile.  Can you try with removing the

**Comment by developeer** [comment]
will do.

**Comment by developeer** [comment]
it is running now.  the analyzer is getting this warning.  Looks like a problem to me.

│ 2019-07-18T13:38:13.830Z    WARN    k8s/probe.go:189 logOnError    skydive-analyzer-6fdc988d6f-nd5hd: github.com/skydive-project/skydive/topology/probes/k │
│ 8s/cache.go:95: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:kube-system:skydive-service-account" cannot list resource "pods" in │
│  API group "" at the cluster scope

**Comment by safchain** [comment]
@hunchback can you please help on this ?

**Comment by safchain** [comment]
It seems to be a permission issue, like Skydive not allowed to access the kubernetes API. I think there a cluster configuration do apply in order to allow Skydive accessing.

You can also disable the Kubernetes probe removing this line
https://github.com/skydive-project/skydive/blob/master/contrib/kubernetes/skydive.yaml#L52

It will anyway get the container detected with the k8s labels but without the k8s objects.

**Comment by hunchback** [comment]
@developeer - please check a workaround of deploying to kube-system namespace - if that works we will then try a more grained fix (using rbac):

```
kubectl apply -f skydive.yaml -n kube-system
```

**Comment by developeer** [comment]
I ran into the same problem when I deployed with the command above.  In order to get the agents to run, I had to change the clusterrole to cluster-admin instead of "view" and remove the readiness and liveness checks on the deployment and daemonset.  It seems the "view" cluster role does not have enough permissions.  Also it seems weird to disable the readiness and liveness probes.

**Comment by hunchback** [comment]
@developeer - I have created a PR will fix for cluster-admin - https://github.com/skydive-project/skydive/pull/1954

Regarding liveness/readiness - can you please provide details on failure

**Comment by chirauki** [comment]
I just raised  #1969 

Binding the agent to 0.0.0.0 did the trick.

**Comment by hunchback** [comment]
merged #1969 ... you can close issue
